README.md
[![Build Status](https://travis-ci.org/mlfmonde/cluster_cli.svg?branch=master)](https://travis-ci.org/mlfmonde/cluster_cli)
[![Coverage Status](https://coveralls.io/repos/github/mlfmonde/cluster_cli/badge.svg?branch=master)](https://coveralls.io/github/mlfmonde/cluster_cli?branch=master)

# Cluster cli

This is a command line utility tool to helps administrator to manage their
[cluster](https://github.com/mlfmonde/cluster).

Features are:

* List consul health checks per nodes and service
* Deploy or switch services
* Migrate anybox/buttervolume docker volumes from one service to an other
* Clear a node by moving all services running on it
* Inspect a node to display all master services of that node
* Run anywhere you can contact your consul API

## Commands

Use the embedded help command, to know the list of available commands:

```bash
$ cluster -h
usage: cluster [-h] [--consul CONSUL] [-y] [-f LOGGING_FILE]
               [-l LOGGING_LEVEL] [--logging-format LOGGING_FORMAT]
               {checks,deploy,migrate,move-masters-from,inspect} ...

Command line utility to administrate cluster

positional arguments:
  {checks,deploy,migrate,move-masters-from,inspect}
                        sub-commands
    checks              List consul health checks per nodes/service
    deploy              Deploy or re-deploy a service
    migrate             Migrate buttervolume (docker volume) data from a
                        service to another one. Only identical volume name
                        will be restored. ``source`` service is where to get
                        data to restore on the ``target`` service. Make sure
                        to snaphot or backup your data (ie: switch your app
                        before migrate) on the target, target data will be
                        lost.``prod`` branch name as a target is forbidden
    move-masters-from   If you want to do some maintenance operation on the
                        host server.This command will helps you to send all
                        events to serviceshosted on the given node to its
                        slave or the wished master
    inspect             Display all master services of given node.

optional arguments:
  -h, --help            show this help message and exit
  --consul CONSUL, -c CONSUL
                        consul api url
  -y, --assume-yes      Always answers ``yes`` to any questions.

Logging params:
  -f LOGGING_FILE, --logging-file LOGGING_FILE
                        Logging configuration file, (logging-level and
                        logging-format are ignored if provide)
  -l LOGGING_LEVEL, --logging-level LOGGING_LEVEL
  --logging-format LOGGING_FORMAT
```

The default consul value is ``http://localhost:8500`` so you may want to
create a ssh tunnel to access to your consul before running this client

```bash
ssh -L 8500:localhost:8500 consul.host.org
```

### Checks

List [consul health checks](https://www.consul.io/api/health.html) per nodes
and service.

```bash
$ cluster checks -h
usage: cluster checks [-h] [-a]

optional arguments:
  -h, --help  show this help message and exit
  -a, --all   Display all checks (any states)
```

_Usage example:_
```bash
$ cluster checks
Node node-3
 - Service ABC
     - Cehck (critical): Service 'ABC' check
```

### Deploy

Re-deploy a service.

if master / slave are not provided, system will choose for you. For replicate
service it will try to switch master/slave. You can force master or slave only.


```bash
$ cluster deploy -h
usage: cluster deploy [-h] [--master NODE] [--slave NODE] [-u] [-d]
                      [-t TIMEOUT]
                      repo branch

positional arguments:
  repo                  The repo name or whole form
                        (ssh://git@git.example.com:22/project-slug/repo-name)
                        for new service
  branch                The branch to deploy

optional arguments:
  -h, --help            show this help message and exit
  --master NODE         Node where to deploy the master (required for new
                        service)
  --slave NODE          Slave node
  -u, --update          Ask for update (update script) before services are up
  -d, --no-wait         Run the script in detached mode : do not wait the end
                        of deployment to stop the script.
  -t TIMEOUT, --timeout TIMEOUT
                        Time in second to let a chance to deploy the service
                        beforeraising an exception (ignored with ``--no-wait``
                        option)
```
### Migrate

Migrate buttervolume (docker volume) data from a service to another one.

Only identical volume name will be restored.

``source`` service is where to get data to restore on the ``target`` service.

Make sure to have a recent snaphot or backup of your target data (ie: switch
your app before migrate) target data will be lost.

``prod`` branch name as a target is forbidden.


```bash
$ cluster migrate -h
usage: cluster migrate [-h] [--target-repo TARGET_REPO] [-d] [-t TIMEOUT]
                       source_repo source_branch target_branch

positional arguments:
  source_repo           The source repo where to get data that will be
                        restored on the target service
  source_branch         The source branch where to get data that will be
                        restored on the target service
  target_branch         The target branch where data will be restored

optional arguments:
  -h, --help            show this help message and exit
  --target-repo TARGET_REPO
                        The target repo if different to the source-repo where
                        data will be restored.
  -d, --no-wait         Run the script in detached mode : do not wait the end
                        of deployment to stop the script.
  -t TIMEOUT, --timeout TIMEOUT
                        Time in second to let a chance to deploy the service
                        beforeraising an exception (ignored with ``--no-wait``
                        option)
  -n, --no-update       When migrate, do not run the upgrade script.
                        Recommended when you migrate a prod db on an iso
                        staging app.
```

### Move from master

This script allow to move all masters hosted on the given node away.

```bash
$ cluster move-masters-from -h
usage: cluster move-masters-from [-h] [-m MASTER] [-d] [-t TIMEOUT] node

positional arguments:
  node                  Node where services should not be hosted that we want
                        to move away

optional arguments:
  -h, --help            show this help message and exit
  -m MASTER, --master MASTER
                        Node to use if no replicate (slave) define on a
                        service, otherwise slave will be used as master.
  -d, --no-wait         Run the script in detached mode : do not wait the end
                        of deployment to stop the script.
  -t TIMEOUT, --timeout TIMEOUT
                        Time in second to let a chance to deploy the service
                        beforeraising an exception (ignored with ``--no-wait``
                        option)
```

### Inspect

Display all master services of given node.

```bash
$ cluster inspect --help
usage: cluster inspect [-h] node

positional arguments:
  node        Node where services should be inspected.

optional arguments:
  -h, --help  show this help message and exit
```

## Install

This tool is tested on python 3.5 ans greater

### Using docker image

> **Note**: The default ``--consul`` parameter is ``http://localhost:8500``
> so you may want to create a ssh tunnel to access to your consul API
> endpoint before running this client

```bash
docker run -it --rm --network host mlfmonde/cluster_cli -h
```

### On you hosted python to use it

```bash
$ pip3 install git+https://github.com/mlfmonde/cluster_cli@master
```

or

```bash
$ git clone https://github.com/mlfmonde/cluster_cli
$ cd cluster_cli
$ pip3 install .
```

### For development with a python virtualenv

```bash
$ python3 -m venv cluster_cli_venv
$ source cluster_cli_venv/bin/activate
$ git clone https://github.com/mlfmonde/cluster_cli
$ cd cluster_cli
$ pip install -r requirements.txt
$ pip install -r requirements.tests.txt
$ python setup.py develop
$ py.test --pep8 --cov=cluster --cov-report=html --lf --nf --ff -v
```


## TODOs

* Find a way to add some checks to avoid deploy some branch (likes qualif) into
  some nodes (reserved for production). this could be implemented by [cluster](
  https://github.com/mlfmonde/cluister) as well


CHANGELOG.rst
Change log
==========

Unreleased
----------

* wait the end of consumption before stop (replace --wait option per --no-wait)

* Build docker image

* Improve various user feedback/usage, add confirmation, logging config and typo

* Implement 'move-masters-from' subcommand to move all services hosted on the given node

* Implement 'migrate' subcommand to move buttervolume docker volumes from a
  service to another one (ie: prod to qualif)

* Implement 'deploy' subcommand to Deploy or re-deploy a service

* First implementation, list checks per nodes

* Implement 'inspect' subcommand to display master apps of a given node


cluster/tests/test_util.py
import unittest
from cluster import util


class TestUtils(unittest.TestCase):

    def test_json2object(self):
        obj = util.json2obj(
            """{
                "dict_var": {
                    "dict-var2": {
                        "string.var": "a string",
                        "int_var": 123,
                        "string-list": [
                            "el1",
                            "el2",
                            "el3"
                        ]
                    }
                },
                "int-array": [
                    1, 2, 3
                ]
            }"""
        )
        self.assertEqual(
            obj.dict_var.dict_var2.string_var,
            "a string"
        )
        self.assertEqual(
            obj.dict_var.dict_var2.int_var,
            123
        )
        self.assertEqual(
            obj.dict_var.dict_var2.string_list,
            ["el1", "el2", "el3"]
        )
        self.assertEqual(
            obj.int_array,
            [1, 2, 3]
        )

    def test_json2object_no_data(self):
        self.assertFalse(util.json2obj(None))


cluster/tests/test_nodes.py
from cluster.cluster import Cluster
from cluster.tests.cluster_test_case import ClusterTestCase


class TestDeploy(ClusterTestCase):

    def test_nodes(self):
        self.cluster = Cluster()
        self.assertEqual(
            self.cluster.nodes,
            ['node-1', 'node-2', 'node-3', 'node-4']
        )


cluster/tests/test_checks.py
from testfixtures import OutputCapture
from unittest import mock

from cluster.client import main
from cluster.tests.cluster_test_case import ClusterTestCase


class TestChecks(ClusterTestCase):

    def test_checks_empty_result(self):
        self.assertFalse(
            self.cluster.checks()
        )

    def fill_data(self):
        def consul_health_state(state):
            if state == 'passing':
                return [
                    {
                        'Node': 'node-1',
                        'ServiceID': 'service-1',
                        'ServiceName': 'Service 1',
                        'Status': state,
                        'Output': "check output",
                        'Name': "Check Service 1",
                    }, {
                        'Node': 'node-2',
                        'ServiceID': 'service-2',
                        'ServiceName': 'Service 2',
                        'Status': state,
                        'Output': "check output",
                        'Name': "Check Service 2",
                    }, {
                        'Node': 'node-2',
                        'ServiceID': 'service-2',
                        'ServiceName': 'Service 2',
                        'Status': state,
                        'Output': "check output 2",
                        'Name': "Check Service 2.2",
                    }
                ]
            elif state == 'critical':
                return [
                    {
                        'Node': 'node-2',
                        'ServiceID': 'service-3',
                        'ServiceName': 'Service 3',
                        'Status': state,
                        'Output': "check output error",
                        'Name': "Check Service 3",
                    }
                ]
            else:
                return []
        self.mocked_consul.configure_mock(**{
            'health.state.side_effect': consul_health_state,
        })

    def test_checks_all(self):
        self.fill_data()
        self.maxDiff = None
        self.assertEqual(
            self.cluster.checks(all=True),
            {
                'node-1': {
                    'service-1': {
                        'name': "Service 1",
                        'checks': [
                            ('Check Service 1', 'passing', 'check output', ),
                        ],
                    },
                },
                'node-2': {
                    'service-2': {
                        'name': "Service 2",
                        'checks': [
                            ('Check Service 2', 'passing', 'check output', ),
                            (
                                'Check Service 2.2',
                                'passing',
                                'check output 2',
                            ),
                        ],
                    },
                    'service-3': {
                        'name': "Service 3",
                        'checks': [
                            (
                                'Check Service 3',
                                'critical',
                                'check output error',
                            ),
                        ],
                    },
                },
            }
        )

    def test_checks_warn(self):
        self.fill_data()
        self.assertEqual(
            self.cluster.checks(all=False),
            {
                'node-2': {
                    'service-3': {
                        'name': "Service 3",
                        'checks': [
                            (
                                'Check Service 3',
                                'critical',
                                'check output error',
                            ),
                        ],
                    },
                },
            }
        )

    def test_check_command_lines(self):
        self.fill_data()
        with OutputCapture() as output:
            with mock.patch('sys.argv', ['cluster', 'checks', ]):
                main()
                output.compare(
                    "\n".join([
                        "Node node-2:",
                        " - Service Service 3:",
                        "    - Check (critical): Check Service 3",
                    ])
                )


cluster/client.py
import argparse
import json
import logging

from cluster import cluster


def main():
    parser = argparse.ArgumentParser(
        prog="cluster",
        description="Command line utility to administrate cluster",
    )
    parser.add_argument(
        '--consul', '-c',
        help="consul api url",
        default="http://localhost:8500"
    )
    parser.add_argument(
        '-y', '--assume-yes', action='store_true',
        help="Always answers ``yes`` to any questions."
    )
    logging_group = parser.add_argument_group(
        'Logging params'
    )
    logging_group.add_argument(
        '-f',
        '--logging-file',
        type=argparse.FileType('r'),
        help='Logging configuration file, (logging-level and logging-format '
             'are ignored if provide)'
    )
    logging_group.add_argument(
        '-l', '--logging-level', default='WARN'
    )
    logging_group.add_argument(
        '--logging-format',
        default='%(asctime)s - %(levelname)s (%(module)s%(funcName)s): '
                '%(message)s'
    )
    subparsers = parser.add_subparsers(help='sub-commands')

    parser_checks = subparsers.add_parser(
        'checks', help='List consul health checks per nodes/service'
    )
    parser_checks.add_argument(
        '-a', '--all', action='store_true',
        help='Display all checks (any states)'
    )

    parser_deploy = subparsers.add_parser(
        'deploy', help='Deploy or re-deploy a service'
    )
    parser_deploy.add_argument(
        'repo',
        help='The repo name or whole form ('
             'ssh://git@git.example.com:22/project-slug/repo-name) '
             'for new service'
    )
    parser_deploy.add_argument(
        'branch',
        help='The branch to deploy'
    )
    parser_deploy.add_argument(
        '--master',
        metavar='NODE',
        help='Node where to deploy the master (required for new service)'
    )
    parser_deploy.add_argument(
        '--slave',
        metavar='NODE',
        help='Slave node'
    )
    parser_deploy.add_argument(
        '-u', '--update',
        action='store_true',
        help='Ask for update (update script) before services are up'
    )

    parser_deploy.add_argument(
        '-d', '--no-wait',
        action='store_true',
        help='Run the script in detached mode : do not wait the end of '
             'deployment to stop the script.'
    )
    parser_deploy.add_argument(
        '-t', '--timeout',
        type=int,
        default=cluster.DEFAULT_TIMEOUT,
        help='Time in second to let a chance to deploy the service before'
             'raising an exception (ignored with ``--no-wait`` option)'
    )

    parser_migrate = subparsers.add_parser(
        'migrate',
        help='Migrate buttervolume (docker volume) data from a service '
             'to another one. Only identical volume name will be restored. '
             '``source`` service is where to get data to restore on the '
             '``target`` service. Make sure to snaphot or backup your data '
             '(ie: switch your app before migrate) on the target, target data '
             'will be lost.'
             '``prod`` branch name as a target is forbidden'
    )
    parser_migrate.add_argument(
        'source_repo',
        help='The source repo where to get data that will be restored on '
             'the target service'
    )
    parser_migrate.add_argument(
        'source_branch',
        help='The source branch where to get data that will be restored on '
             'the target service'
    )
    parser_migrate.add_argument(
        'target_branch',
        help='The target branch where data will be restored'
    )
    parser_migrate.add_argument(
        '--target-repo',
        help='The target repo if different to the source-repo where data '
             'will be restored.'
    )
    parser_migrate.add_argument(
        '-d', '--no-wait',
        action='store_true',
        help='Run the script in detached mode : do not wait the end of '
             'deployment to stop the script.'
    )
    parser_migrate.add_argument(
        '-t', '--timeout',
        type=int,
        default=cluster.DEFAULT_TIMEOUT,
        help='Time in second to let a chance to deploy the service before '
             'raising an exception (ignored with ``--no-wait`` option)'
    )
    parser_migrate.add_argument(
        '-n', '--no-update',
        action='store_true',
        default=False,
        help='When migrate, do not run the upgrade script. '
             'Recommended when you migrate a prod db on an iso staging app.'
    )

    parser_move_masters_from = subparsers.add_parser(
        'move-masters-from',
        help='If you want to do some maintenance operation on the host server.'
             'This command will helps you to send all events to services'
             'hosted on the given node to its slave or the wished master'
    )
    parser_move_masters_from.add_argument(
        'node',
        help='Node where services should not be hosted '
             'that we want to move away'
    )
    parser_move_masters_from.add_argument(
        '-m', '--master',
        help="Node to use if no replicate (slave) define on a service, "
             "otherwise slave will be used as master."
    )
    parser_move_masters_from.add_argument(
        '-d', '--no-wait',
        action='store_true',
        help='Run the script in detached mode : do not wait the end of '
             'deployment to stop the script.'
    )
    parser_move_masters_from.add_argument(
        '-t', '--timeout',
        type=int,
        default=cluster.DEFAULT_TIMEOUT,
        help='Time in second to let a chance to deploy the service before'
             'raising an exception (ignored with ``--no-wait`` option)'
    )

    parser_inspect = subparsers.add_parser(
        'inspect',
        help='Display all master services of given node.'
    )
    parser_inspect.add_argument(
        'node',
        help='Node where services should be inspected.'
    )

    def init(args):
        return cluster.Cluster(
            args.consul
        )

    def cluster_checks(args):
        cluster = init(args)
        for node, services in cluster.checks(all=args.all).items():
            print("Node {}:".format(node))
            for _, service in services.items():
                print(" - Service {}:".format(service['name']))
                for name, status, _ in service['checks']:
                    print("    - Check ({}): {}".format(status, name))

    def cluster_deploy(args):
        cluster = init(args)
        cluster.deploy(
            args.repo,
            args.branch,
            master=args.master,
            slave=args.slave,
            no_wait=args.no_wait,
            timeout=args.timeout,
            ask_user=not args.assume_yes,
            update=args.update
        )

    def cluster_migrate(cmd_args):
        cluster = init(cmd_args)
        cluster.migrate(
            cmd_args.source_repo,
            cmd_args.source_branch,
            cmd_args.target_branch,
            target_repo=cmd_args.target_repo,
            no_wait=cmd_args.no_wait,
            timeout=cmd_args.timeout,
            ask_user=not arguments.assume_yes,
            no_update=cmd_args.no_update
        )

    def cluster_move_masters_from(args):
        cluster = init(args)
        cluster.move_masters_from(
            args.node,
            master=args.master,
            no_wait=args.no_wait,
            timeout=args.timeout,
            ask_user=not args.assume_yes
        )

    def cluster_inspect(args):
        cluster = init(args)
        cluster.inspect_node(args.node)

    parser_checks.set_defaults(func=cluster_checks)
    parser_deploy.set_defaults(func=cluster_deploy)
    parser_migrate.set_defaults(func=cluster_migrate)
    parser_move_masters_from.set_defaults(func=cluster_move_masters_from)
    parser_inspect.set_defaults(func=cluster_inspect)

    arguments = parser.parse_args()

    logging.basicConfig(
        level=getattr(logging, arguments.logging_level.upper()),
        format=arguments.logging_format
    )
    if arguments.logging_file:
        try:
            json_config = json.loads(arguments.logging_file.read())
            logging.config.dictConfig(json_config)
        except json.JSONDecodeError:
            logging.config.fileConfig(arguments.logging_file.name)

    if hasattr(arguments, 'func'):
        arguments.func(arguments)
    else:
        parser.print_help()
